mp4parse/src/unstable.rs

```diff
@@ -0,0 +1,547 @@
+    type Item = (u32, u32);
+
+    fn next(&mut self) -> Option<(u32, u32)> {
+        let has_chunk = self.chunks.next().or_else(|| {
+            self.chunks = self.locate();
+            self.remain_chunk_count
+                .checked_sub(
+                    self.chunks
+                        .len()
+                        .try_into()
+                        .expect("len() of a Range<u32> must fit in u32"),
+                )
+                .and_then(|res| {
+                    self.remain_chunk_count = res;
+                    self.chunks.next()
+                })
+        });
+
+        has_chunk.map(|id| (id, self.sample_count))
+    }
+}
+
+impl<'a> SampleToChunkIterator<'a> {
+    #[allow(clippy::reversed_empty_ranges)]
+    fn locate(&mut self) -> std::ops::Range<u32> {
+        loop {
+            return match (self.stsc_peek_iter.next(), self.stsc_peek_iter.peek()) {
+                (Some(next), Some(peek)) if next.first_chunk == peek.first_chunk => {
+                    // Invalid entry, skip it and will continue searching at
+                    // next loop iteration.
+                    continue;
+                }
+                (Some(next), Some(peek)) if next.first_chunk > 0 && peek.first_chunk > 0 => {
+                    self.sample_count = next.samples_per_chunk;
+                    (next.first_chunk - 1)..(peek.first_chunk - 1)
+                }
+                (Some(next), None) if next.first_chunk > 0 => {
+                    self.sample_count = next.samples_per_chunk;
+                    // Total chunk number in 'stsc' could be different to 'stco',
+                    // there could be more chunks at the last 'stsc' record.
+                    match next.first_chunk.checked_add(self.remain_chunk_count) {
+                        Some(r) => (next.first_chunk - 1)..r - 1,
+                        _ => 0..0,
+                    }
+                }
+                _ => 0..0,
+            };
+        }
+    }
+}
+
+/// Calculate numerator * scale / denominator, if possible.
+///
+/// Applying the associativity of integer arithmetic, we divide first
+/// and add the remainder after multiplying each term separately
+/// to preserve precision while leaving more headroom. That is,
+/// (n * s) / d is split into floor(n / d) * s + (n % d) * s / d.
+///
+/// Return None on overflow or if the denominator is zero.
+fn rational_scale<T, S>(numerator: T, denominator: T, scale2: S) -> Option<T>
+where
+    T: PrimInt + Zero,
+    S: PrimInt,
+{
+    if denominator.is_zero() {
+        return None;
+    }
+
+    let integer = numerator / denominator;
+    let remainder = numerator % denominator;
+    num_traits::cast(scale2).and_then(|s| match integer.checked_mul(&s) {
+        Some(integer) => remainder
+            .checked_mul(&s)
+            .and_then(|remainder| (remainder / denominator).checked_add(&integer)),
+        None => None,
+    })
+}
+
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord)]
```

Comment:
Debug and PartialEq are needed for tests, otherwise this fails to compile: 
```
assert_eq!(
    track_time_to_us(duration, scale),
    Some(Microseconds(100_079_991_719_000_000u64))
);
```